apps/src/code-studio/components/header/HeaderMiddle.jsx

```diff
@@ -0,0 +1,309 @@
+import _ from 'lodash';
+import $ from 'jquery';
+
+const styles = {
+  headerMiddleContent: {
+    width: '100%',
+    display: 'flex',
+    alignItems: 'center',
+    animation: 'header_fadein 0.4s'
+  },
+  finishedLink: {
+    width: '100%'
+  }
+};
+
+class HeaderMiddle extends React.Component {
+  static propTypes = {
+    projectInfoOnly: PropTypes.bool,
+    appLoadStarted: PropTypes.bool,
+    appLoaded: PropTypes.bool,
+    scriptNameData: PropTypes.object,
+    lessonData: PropTypes.object,
+    lessonExtrasUrl: PropTypes.string,
+    scriptData: PropTypes.object,
+    currentLevelId: PropTypes.string,
+    linesOfCodeText: PropTypes.string
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      width: this.getWidth(),
+      projectInfoDesiredWidth: 0,
+      scriptNameDesiredWidth: 0,
+      lessonProgressDesiredWidth: 0,
+      finishDesiredWidth: 0,
+      initialDelay: true
+    };
+
+    setTimeout(() => {
+      this.setState({initialDelay: false});
+    }, 750);
+  }
+
+  componentDidMount() {
+    this.updateLayout();
+
+    this.updateLayoutListener = _.throttle(this.updateLayout, 200);
+    window.addEventListener('resize', this.updateLayoutListener);
+    window.addEventListener('scroll', this.updateLayoutListener);
+  }
+
+  getWidth() {
+    const width = $('.header_middle').width();
+    return width;
+  }
+
+  updateLayout = () => {
+    this.setState({
+      width: this.getWidth(),
+      windowWidth: $(window).width(),
+      windowHeight: $(window).height()
+    });
+  };
+
+  setDesiredWidth(componentName, width) {
+    const stateComponentName = componentName + 'DesiredWidth';
+
+    if (Math.ceil(width) !== this.state[stateComponentName]) {
+      this.setState({[stateComponentName]: Math.ceil(width)});
+    }
+  }
+
+  // Return the desired widths for the items that are showing.
+  // Also return whether we are only showing the HeaderPopup because we are
+  // cropping the progress, and otherwise wouldn't show it.
+  getWidths() {
+    const width = this.state.width;
+
+    if (this.props.projectInfoOnly) {
+      return {
+        projectInfo: Math.floor(
+          Math.min(this.state.projectInfoDesiredWidth, width)
+        ),
+        scriptName: 0,
+        progress: 0,
+        popup: 0,
+        finish: 0
+      };
+    }
+
+    const lessonProgresssDesiredWidth =
+      this.state.lessonProgressDesiredWidth + 10;
+
+    const showFinish = !!(
+      this.props.lessonData && this.props.lessonData.finishLink
+    );
+
+    // projectInfo gets no more than 30% of the entire width
+    const projectInfoWidth = Math.floor(
+      Math.min(this.state.projectInfoDesiredWidth, width * 0.3)
+    );
+
+    let remainingWidth = width - projectInfoWidth;
+
+    // progress gets no more than 60% of the remaining width
+    const progressWidth = Math.floor(
+      Math.min(lessonProgresssDesiredWidth, remainingWidth * 0.6)
+    );
+
+    remainingWidth = remainingWidth - progressWidth;
```

Comment:
Can you add a few more comments to the sections here to help split them into logical sections for the reader?